dataset/src/main/scala/frameless/macros/ColumnMacros.scala

```diff
@@ -0,0 +1,377 @@
+
+      // A unary operator - the operator must exist on org.apache.spark.sql.Column
+      case sel @ Select(This(rhsE), op: TermName) => Some {
+        subUnaryOperators.get(sel.symbol.asMethod)
+          .map {
+            sub => rhsE.right.map(rhs => TrustMeBro(tree.tpe, sub(Expr.refold(rhs))))
+          }.getOrElse {
+            if(isColumnOp(op, 0)) {
+              rhsE.right.map {
+                rhs => UnaryRAOperator(tree.tpe, rhs, op)
+              }
+            } else addError(rhsE)(tree, s"${op.decodedName} is not a valid column operator")
+          }
+      }
+
+      // A unary operator (which is a no-arg method) - the operator must exist on org.apache.spark.sql.Column
+      case Apply(sel @ Select(This(rhsE), op: TermName), List()) => Some {
+        subUnaryOperators.get(sel.symbol.asMethod)
+          .map {
+            sub => rhsE.right.map(rhs => TrustMeBro(tree.tpe, sub(Expr.refold(rhs))))
+          }.getOrElse {
+
+          if(isColumnOp(op, 0)) {
+            // $COVERAGE-OFF$ - can't find a unary column op that isn't substituted and is a no-arg apply tree
+            rhsE.right.map {
+              rhs => UnaryRAOperator(tree.tpe, rhs, op)
+            }
+            // $COVERAGE-ON$
+          } else addError(rhsE)(tree, s"${op.decodedName} is not a valid column operator")
+        }
+      }
+
+      // A literal constant (would it be useful to distinguish this from non-constant literal?
+      case Literal(_) => Some(Right(LiteralExpr(tree.tpe, tree)))
+
+      // Constructing a case class
+      case Apply(sel @ Select(qualifier, TermName("apply")), AllExprs(argsE)) => Some {
+        if (isConstructor(sel, tree.tpe)) {
+          argsE.right.map {
+            args =>
+              val params = sel.symbol.asMethod.paramLists.head
+              val names = sel.symbol.asMethod.paramLists.head.zip(args).map {
+                case (param, paramExpr) => param.name.toString -> paramExpr
+              }.toMap
+              Construct(tree.tpe, names)
+          }
+        } else addError(argsE)(tree, "Only constructor can be used here")
+      }
+      // Constructing a tuple or parameterized case class
+      case Apply(ta @ TypeApply(sel @ Select(qualifier, TermName("apply")), typArgs), AllExprs(argsE)) => Some {
+        if(isConstructor(ta, tree.tpe)) {
+          argsE.right.map {
+            args =>
+              val params = sel.symbol.asMethod.paramLists.head
+              val names = sel.symbol.asMethod.paramLists.head.zip(args).map {
+                case (param, paramExpr) => param.name.toString -> paramExpr
+              }.toMap
+              Construct(tree.tpe, names)
+          }
+        } else addError(argsE)(tree, "Only constructor can be used here")
+      }
+
+      // A binary operator - the operator must exist on org.apache.spark.sql.Column
+      case Apply(sel @ Select(This(lhsE), op: TermName), List(This(rhsE))) => Some {
```

Comment:
Maybe typechecking the operation against a `TypedColumn` tree would be better here? But we would have to implement all the operators from `Column` in `TypedColumn` and I think a lot are missing